Source/WebCore/Modules/permissions/Permissions.cpp

```diff
@@ -77,66 +82,81 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
-        });
+    if (document && !isAllowedByFeaturePolicy(*document, permissionDescriptor.name)) {
+        promise.resolve(PermissionStatus::create(*context, PermissionState::Denied, permissionDescriptor));
         return;
     }
 
     auto* origin = context->securityOrigin();
     auto originData = origin ? origin->data() : SecurityOriginData { };
 
-    PermissionController::shared().query(ClientOrigin { context->topOrigin().data(), originData }, PermissionDescriptor { parameterDescriptor }, *document->page(), [this, protectedThis = Ref { *this }, parameterDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
-        auto context = m_navigator ? m_navigator->scriptExecutionContext() : nullptr;
-        if (!context || !context->globalObject())
-            return;
+    if (auto* workerGlobalScope = dynamicDowncast<WorkerGlobalScope>(context)) {
+        auto completionHandler = [originData = WTFMove(originData).isolatedCopy(), permissionDescriptor, contextIdentifier = workerGlobalScope->identifier(), promise = WTFMove(promise)] (auto& context) mutable { 
+            ASSERT(isMainThread());
 
-        context->postTask([parameterDescriptor, promise = WTFMove(promise),  permissionState = WTFMove(permissionState)](auto& context) mutable {
-            if (!permissionState) {
-                promise.reject(Exception { NotSupportedError });
+            auto* document = downcast<Document>(&context);
+            if (!document->page()) {
+                promise.reject(Exception { InvalidStateError, "The page does not exist"_s });
                 return;
             }
-            promise.resolve(PermissionStatus::create(context, *permissionState, parameterDescriptor));
-        });
+
+            PermissionController::shared().query(ClientOrigin { context.topOrigin().data(), WTFMove(originData) }, PermissionDescriptor { permissionDescriptor }, *document->page(), [context = Ref { context }, contextIdentifier, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
+                if (!context.get().globalObject()) {
+                    promise.reject(Exception { InvalidStateError, "The context is invalid"_s });
+                    return;
+                }
+
+                ScriptExecutionContext::postTaskTo(contextIdentifier, [promise = WTFMove(promise), permissionState, permissionDescriptor](auto& context) mutable {
+                    if (!permissionState)
+                        promise.reject(Exception { NotSupportedError, "Permissions::query does not support this API"_s });
+                    else
+                        promise.resolve(PermissionStatus::create(context, *permissionState, permissionDescriptor));
+                });
+            });
+        };
+
+        workerGlobalScope->thread().workerLoaderProxy().postTaskToLoader(WTFMove(completionHandler));
+        return;
+    }
+
```

Comment:
Please `ASSERT(document);` here and use `document` instead of `context` below.

Reply:
You're right--we could just ref Document, and capture that. But then would we remove the `!context->globalObject()` check below? If it can be removed, then maybe it can be removed from the start of the function as well? (on line 86, where we are going to remove the unneeded null check). Also, can document be used instead of context to create the `PermissionStatus`?

Reply:
You can check `!document->globalObject()`, a document is a context.

Reply:
Ah got it--and we can check `!document->globalObject()` even before the lambda right? It would be clean design to do the check in the part of the code where we have other checks relating to the document (line 92 'if (document) ...' ).

Reply:
Before the lambda is probably not a good idea since the lambda runs asynchronously. It is best to check the globalObject right before resolving the promise.

Reply:
Ok sounds good. Do we prefer rejecting the promise if the globalObject check fails or it is ok to use an assert for the globalObject?. Because we expect the context to be a globalObject right?

Reply:
`ASSERT(document);
    PermissionController::shared().query( ... [context = Ref { context }, ...](...) mutable {
        ASSERT(context.get()->globalObject());

        if (!permissionState)
            promise.reject(Exception { NotSupportedError, "Permissions::query does not support this API"_s });
        else
            promise.resolve(PermissionStatus::create(*document, *permissionState, permissionDescriptor));
    });`

Reply:
Like this:
<img width="787" alt="image" src="https://user-images.githubusercontent.com/42721308/184281178-452516eb-2ba7-4e5d-936f-1c3a2f727097.png">

Reply:
Turns out the `globalObject` test is only needed for getting `vm`, so the check does not need to happen anywhere else. Fixed!